kotlin/android/app/src/main/java/dev/firezone/android/tunnel/NetworkMonitor.kt

```diff
@@ -0,0 +1,21 @@
+/* Licensed under Apache 2.0 (C) 2024 Firezone, Inc. */
+import android.net.ConnectivityManager
+import android.net.LinkProperties
+import android.net.Network
+import com.google.gson.Gson
+import com.squareup.moshi.Moshi
+import dev.firezone.android.tunnel.ConnlibSession
+import javax.inject.Inject
+
+class NetworkMonitor(private val connlibSessionPtr: Long) : ConnectivityManager.NetworkCallback() {
+    @Inject
+    internal lateinit var moshi: Moshi
+
+    override fun onLinkPropertiesChanged(
+        network: Network,
+        linkProperties: LinkProperties,
+    ) {
+        ConnlibSession.networkUpdate(connlibSessionPtr, Gson().toJson(linkProperties.dnsServers))
```

Comment:
```suggestion
        ConnlibSession.networkUpdate(connlibSessionPtr, moshi.adapter<MutableList<String>>().fromJson(dnsAddresses)!!)
```

https://github.com/firezone/firezone/blob/main/kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt#L105

Reply:
`dnsAddresses` isn't in json format

Reply:
If we want to use `moshi` to serialize the link properties list we need to create a new wrapper type that implements the encoding, `gson` offers this easier way to do it.